server/src/main/java/com/tencent/rss/server/ShuffleServerGrpcService.java

```diff
@@ -426,21 +427,68 @@ public void getShuffleData(GetShuffleDataRequest request,
     responseObserver.onCompleted();
   }
 
-  private List<ShuffleDataBlockSegment> toBlockSegments(List<BufferSegment> bufferSegments) {
-    List<ShuffleDataBlockSegment> ret = Lists.newArrayList();
-
-    for (BufferSegment segment : bufferSegments) {
-      ret.add(ShuffleDataBlockSegment.newBuilder()
-          .setBlockId(segment.getBlockId())
-          .setOffset(segment.getOffset())
-          .setLength(segment.getLength())
-          .setUncompressLength(segment.getUncompressLength())
-          .setCrc(segment.getCrc())
-          .setTaskAttemptId(segment.getTaskAttemptId())
-          .build());
+  @Override
+  public void getShuffleIndex(GetShuffleIndexRequest request,
+      StreamObserver<GetShuffleIndexResponse> responseObserver) {
+    ShuffleServerMetrics.counterTotalRequest.inc();
+
+    String appId = request.getAppId();
+    int shuffleId = request.getShuffleId();
+    int partitionId = request.getPartitionId();
+    int partitionNumPerRange = request.getPartitionNumPerRange();
+    int partitionNum = request.getPartitionNum();
+    StatusCode status = StatusCode.SUCCESS;
+    String msg = "OK";
+    GetShuffleIndexResponse reply;
+    String requestInfo = "appId[" + appId + "], shuffleId[" + shuffleId + "], partitionId["
+        + partitionId + "]";
+
+    if (shuffleServer.isMultiStorageEnabled()) {
+      shuffleServer.getMultiStorageManager().updateLastReadTs(appId, shuffleId, partitionId);
     }
 
-    return ret;
+    // Use an default file size here because the index file size is very small and not need to get the precise size
+    // from the client side by using the bitmap cardinality nor in the server side by using the read handler.
+    // A index segment has 40B and a block's average size is about 10MB, so an index file with the default size 2MB
+    // could index about 500GB shuffle data and the default size is big enough to hold all the index data in memory.
```

Comment:
Index file is expected small size and won't cause oom problem with the assumed size. An index segment is 40B, with the default size - 2MB, it can support XXXXXX blocks for shuffle data.

Reply:
ok